### PR TITLE
I18n tools

### PR DESCRIPTION
**Pull request by yallxe**
WIP

Planning to add rules translations (or something similar) and mozilla fluent support

### CODE REVIEW

File: telegrinder/tools/i18n/base.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    @abstractmethod
+    def __init__(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_translator_by_locale(self, locale: str) -> "AbstractTranslator":
+        raise NotImplementedError
+
+
+class AbstractTranslator(ABC):
```
Comment (timoniq) [inline telegrinder/tools/i18n/base.py:14]: We use `ABC` instead of `Abstract`

File: telegrinder/tools/i18n/middleware/base.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+
+class ABCTranslatorMiddleware(ABCMiddleware):
+    def __init__(self, i18n: AbstractI18n, kwarg_name: str):
+        self.i18n = i18n
+        self.kwarg_name = kwarg_name
+
+    @abstractmethod
+    async def get_locale(self, event) -> str:
+        raise NotImplementedError
```
Comment (timoniq) [inline telegrinder/tools/i18n/middleware/base.py:14]: this is abstract method, raising exception is redundant

File: telegrinder/tools/i18n/middleware/base.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
+from abc import abstractmethod
+
+from telegrinder.bot.dispatch.middleware import ABCMiddleware
+from telegrinder.tools.i18n import AbstractI18n
+
+
+class ABCTranslatorMiddleware(ABCMiddleware):
+    def __init__(self, i18n: AbstractI18n, kwarg_name: str):
+        self.i18n = i18n
+        self.kwarg_name = kwarg_name
```
Comment (timoniq) [inline telegrinder/tools/i18n/middleware/base.py:10]: looks kinda strange, at least default name needed

**Comment by timoniq** [comment]
Hi! Thank you for PR. I added some feedback comments. Looks good at this point in general

File: telegrinder/tools/i18n/base.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
+from abc import ABC, abstractmethod
+
+
+class ABCI18n(ABC):
+    @abstractmethod
+    def __init__(self):
```
Comment (timoniq) [inline telegrinder/tools/i18n/base.py:6]: it seems wrong
there should be no `__init__` abstract method

File: telegrinder/bot/dispatch/handler/func.py [hunk 5]
```diff
@@ -28,9 +28,10 @@ def __init__(
         self.dataclass = dataclass
         self.ctx = {}
 
-    async def check(self, api: ABCAPI, event: Update) -> bool:
-        preset_ctx = self.ctx
-        self.ctx = preset_ctx.copy()
+    async def check(self, api: ABCAPI, event: Update, ctx: dict = None) -> bool:
```
Comment (timoniq) [inline telegrinder/bot/dispatch/handler/func.py:31]: wtf? `ctx: dict = None`

File: telegrinder/bot/rules/abc.py [hunk 6]
```diff
@@ -86,3 +88,23 @@ class MessageRule(ABCRule[Message], ABC, require=[]):
...
+        self.inner_rule = inner_rule
+        self.require = inner_rule.require  # Not sure if this is correct
+        self.adapter = inner_rule.adapter
+
+    @abstractmethod
+    async def check(self, message: Message, ctx: dict) -> bool:
+        ...
+
+
+class TranslatableRuleMixin(ABC, typing.Generic[InnerRuleT]):
```
Comment (timoniq) [inline telegrinder/bot/rules/abc.py:107]: wtf is it called mixin? translate method should be in `ABCTranslatedRule`

File: telegrinder/bot/rules/text.py [hunk 7]
```diff
@@ -10,14 +11,28 @@ class TextMessageRule(MessageRule, ABC, require=[HasText()]):
...
+        ) in texts
+
+    async def check(self, message: Message, ctx: dict) -> bool:
+        return self.check_text(message.text)
+
+    def translate(self) -> "TranslatedText":
+        return TranslatedText(inner_rule=self)
+
+
+class TranslatedText(ABCTranslatedRule[Text]):
```
Comment (timoniq) [inline telegrinder/bot/rules/text.py:34]: nonsense it is better to just generate a `Text` rule with translated Text.texts

**Review by timoniq — CHANGES_REQUESTED** [review]

File: telegrinder/bot/rules/text.py (hunk 7 above)
Comment (yallxe) [reply telegrinder/bot/rules/text.py:34]: You've probably misunderstood my idea of translated rules. `translate()` doesn't translate texts to specified locale, instead it creates a new rule that takes translator context and translates texts inside the `check()` method. We've discussed more in telegram dm.

Comment (timoniq) [reply telegrinder/bot/rules/text.py:34]: I will redesign it soon as for now it looks just bad

**Review by timoniq — APPROVED** [review]

**Pull request merged**